src/librustc_session/session.rs

```diff
@@ -1021,6 +1021,32 @@ impl Session {
         || self.opts.debugging_opts.sanitizer.intersects(SanitizerSet::ADDRESS | SanitizerSet::MEMORY)
     }
 
+    pub fn link_dead_code(&self) -> bool {
+        match self.opts.cg.link_dead_code {
+            Some(explicitly_set) => explicitly_set,
+            None => {
+                self.opts.debugging_opts.instrument_coverage
+                    && !self.target.target.options.is_like_msvc
+                // Issue #64685: (Linker error because `-wL as-needed` conflicts with
+                // `-C link-dead-code`). As described by this issue, the "link dead code" option
+                // produces incorrect binaries when compiled and linked under MSVC. The resulting
+                // Rust programs typically crash with a segmentation fault, or produce an empty
+                // "*.profraw" file (profiling counter results normally generated during program
+                // exit).
+                //
+                // If not targeting MSVC, `-Z instrument-coverage` implies `-C link-dead-code`, so
+                // unexecuted code is still counted as zero, rather than be optimized out. Note that
+                // instrumenting dead code can be explicitly disabled with:
+                //
+                //     `-Z instrument-coverage -C link-dead-code=no`.
+                //
+                // FIXME(richkadel): Investigate if `instrument-coverage` implementation can inject
+                // [zero counters](https://llvm.org/docs/CoverageMappingFormat.html#counter) in the
+                // coverage map when "dead code" is removed, rather than forcing `link-dead-code`.
```

Comment:
I may have commented on this before but I think we'll need to do this at some point. We're wrestling with the same thing for clang as I understand it.

Reply:
Yes, actually the comment is there because you suggested it! 😄 

The concern I have (now that I've looked at what we do based on the value of `link-dead-code`) is this is only a hint to the linker. The native linker is identifying the "dead code" (which, by the way, may not always be functions... I think this option also affects things like unreachable if or else blocks). I don't know if it's possible or practical to find out what "dead code" was removed in a way that we can update the coverage map. (And by the way, the coverage map is generated before we get to the linker.)

It would be cool to do, and perhaps we actually CAN do this for some code that gets eliminated during a MIR pass.

Reply:
I updated comments to include this observation.

Reply:
Right, we wouldn't have that information before getting to the linker. So we probably need a way to tell the linker to preserve the counters, but not the code associated with them.

Reply:
I'd also suggest opening an issue for design questions like this so a conversation can develop around them.